File: Intersection_of_Two_Arrays.cpp
/*349. Intersection of Two Arrays  QuestionEditorial Solution  My Submissions
Total Accepted: 44741
Total Submissions: 100407
Difficulty: Easy
Given two arrays, write a function to compute their intersection.

Example:
Given nums1 = [1, 2, 2, 1], nums2 = [2, 2], return [2].

Note:
Each element in the result must be unique.
The result can be in any order.
Subscribe to see which companies asked this question*/



/*1. push nums1 into hash set
2. check if element in nums2 appears in the set, after check, if exist remove it from the set*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        unordered_set<int> myset(nums1.begin(),nums1.end());
        vector<int> ret;
        for(int n:nums2){
            if(myset.count(n)){
                ret.push_back(n);
                myset.erase(n);
            }
        }
        
        return ret;
    }
};





/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> myset = new HashSet<>();
        List<Integer> mylist = new ArrayList<>();
        
        for(int n:nums1){
            myset.add(n);
        }
        
        for(int n:nums2){
            if(myset.contains(n)){
                mylist.add(n);
                myset.remove(n);
            }
        }
        
        return mylist.stream().mapToInt(b->b).toArray();
            
    }
}

File: Intersection_of_Two_Arrays.swift
/*
Given two arrays, write a function to compute their intersection.

Example 1:

Input: nums1 = [1,2,2,1], nums2 = [2,2]
Output: [2]
Example 2:

Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
Output: [9,4]
Note:

Each element in the result must be unique.
The result can be in any order.
*/

/*
Solution 1:
put one array into a set
then iteratively search another array element exist or not

Time Complexity: O(n+m), where n is nums1.count, m is nums2.count
Space Compelxity: O(n)
*/
class Solution {
    func intersection(_ nums1: [Int], _ nums2: [Int]) -> [Int] {
        var set1 = Set(nums1)
        var res = [Int]()
        for n in nums2 {
            if set1.contains(n) {
                res.append(n)
                set1.remove(n)
            }
        }
        return res
    }
}